### ISSUE TITLE
Handler of messages ECHO_REPLY

### ISSUE THREAD

**Issue by italovalcy** [noxrepo/pox#91]
Dear all,

I would like to suggest include a handler of OFPT_ECHO_REPLY in events handler of openflow messages in POX. This is important for applications that implements some kind of keepalive for their nodes. I known that this feature of keepalive was already part of POX project, but using the above approach, with a handler for echo-reply messages, is more flexible for applications.

Bellow is a patch propose for this feature.

Thanks for any help

Kind Regards, Italo

File: pox-handle-echo-reply.patch

```
diff --git a/pox/openflow/__init__.py b/pox/openflow/__init__.py
index 9a4d6d3..1930586 100644
--- a/pox/openflow/__init__.py
+++ b/pox/openflow/__init__.py
@@ -241,6 +241,14 @@ class ConnectionIn (Event):
     self.dpid = connection.dpid
     self.nexus = None

+class EchoReply (Event):
+  """
+  Event raised when an switch replies echo request.
+  """
+  def __init__ (self, connection, ofp):
+    Event.__init__(self)
+    self.connection = connection
+    self.dpid = connection.dpid


 class OpenFlowConnectionArbiter (EventMixin):
@@ -314,6 +322,7 @@ class OpenFlowNexus (EventMixin):
     PortStatsReceived,
     QueueStatsReceived,
     FlowRemoved,
+    EchoReply,
   ])

   # Bytes to send to controller when a packet misses all flows
diff --git a/pox/openflow/of_01.py b/pox/openflow/of_01.py
index 4096714..7d82a43 100644
--- a/pox/openflow/of_01.py
+++ b/pox/openflow/of_01.py
@@ -74,7 +74,9 @@ def handle_HELLO (con, msg): #S

 def handle_ECHO_REPLY (con, msg):
   #con.msg("Got echo reply")
-  pass
+  e = con.ofnexus.raiseEventNoErrors(EchoReply, con, msg)
+  if e is None or e.halt != True:
+    con.raiseEventNoErrors(EchoReply, con, msg)

 def handle_ECHO_REQUEST (con, msg): #S
   reply = msg
@@ -587,6 +589,7 @@ class Connection (EventMixin):
     PortStatsReceived,
     QueueStatsReceived,
     FlowRemoved,
+    EchoReply,
   ])

   # Globally unique identifier for the Connection instance
```

**Comment by MurphyMc** [comment]
I'm not sure why this would be useful.  Can you provide an example?

**Comment by italovalcy** [comment]
Dear MurphyMc,

Thanks for the reply. Actually I have two practical examples where I need a application which makes ECHO REQUEST/REPLY (way OFCTL -> Switch):
 1 - I have one application which store information of OF nodes on a graph (DiGraph, from networkx) and I need to keep the consistence between my network and the graph. In this situation I need a keepalive approach, which sends ECHO_REQUEST and update idle_time when receive the REPLY

 2 - Openflow/Openwrt (stanford implementation) seems to need of some kind of keep alive, because the normal checks (switch -> OFCTL) does not work well. Yet more, the classical approach (switch -> OFCTL) does not helps to detect nodes failures.

Did you got the ideas?

Regards, Italo.

**Comment by MurphyMc** [comment]
(This has been discussed some on the pox-dev mailing list, so you may want to see if you can find the discussion in the archive.)

Connections already keep track of their idle time (I believe it's in Connection.idle_time).  This is actually more than just whether they are sending echo replies -- it's whether they are sending anything at all.  This can be significant, because the control channel may just be very active and while the switch is still sending messages, an echo reply may be delayed.  Disconnecting such a switch as being "idle" is neither particularly true, nor particularly useful -- if the control channel was backlogged before, it's just going to get backlogged again as soon as it reconnects.

The openflow.keepalive component uses the idle-time-tracking feature along with sending echo requests to detect failed connections/switches and disconnect them on the POX side.  Admittedly, this component is not named particularly well, but the name stems from exactly the second point you make (that there's a bug in the reference implementation of OpenFlow which causes it to disconnect).

**Comment by italovalcy** [comment]
Hi MurphyMc,

Thanks for the reply! I saw that early versions of POX really made the task of verifying active nodes in the exact way that you mention above: not only replies to request, but any traffic that comes from the node (it used to be implemented on the method that handles connections from the socket layer). However, from the best of my knowledgement, this features seems to be removed from POX (Connection.idle_time doesn't exists anymore) and now the POX controller just don't known of the lost connections and don't raise ConnectionDown on failure node events (only on normal node shutdown).

Another point is that, as far as I could see, openflow.keepalive componet was removed from POX.

In summary, I think that it would be interesting to have an handler callback for ECHO_REPLY messages, or yet more: for any kind of messages, even the vendors specific messages. In this way, the experimenter just register an function handler to handle a specific kind of message (just like we register components).

**Comment by MurphyMc** [comment]
That functionality has not been removed -- you must be looking at an old branch of POX.  If you look at the most recent release or active branches (betta and carp respectively), it should be there.

As for vendor message handling, there's no generic vendor event because, in my opinion, this would be an ungainly way to handle it.  I actually wrote the code for it once and then didn't merge it.  A vendor message should probably have an Event subclass of its own, and should probably have a message type of its own.  The openflow.nicira component shows how one can "hook" the existing mechanism to change both how vendor messages are decoded from the wire as well as what events they raise.  (This has recently been discussed a little on the list.)  The actual mechanism for hooking isn't currently very beautiful, but the result is that without modifying any of the core infrastructure, you can add new message types and new events corresponding to vendor messages.  This is a lot cleaner than a generic vendor message where the message is just a blob of bytes.  Or so I tell myself.

**Comment by italovalcy** [comment]
Hi MurphyMc,

Thanks for your information. I saw my fault: I think I'm using an old version:

```
cat .git/config 
[core]
    repositoryformatversion = 0
    filemode = true
    bare = false
    logallrefupdates = true
[remote "origin"]
    fetch = +refs/heads/*:refs/remotes/origin/*
    url = https://github.com/noxrepo/pox.git
[branch "master"]
    remote = origin
    merge = refs/heads/master
```

I really don't know how I got this.. The new version uses branch "betta" (or "carp").

Thank you very much for the help.

Regards, italo.
